net: lib: wifi_credentials: Use getopt API for adding network options

D-Triveni opened the pull request:
Use getopt API to process arguments for `cred add` command.

D-Triveni committed:
net: lib: wifi_credentials: Use getopt API for adding network options

Use getopt API to process arguments for `cred add` command.

Signed-off-by: Triveni Danda <triveni.danda@nordicsemi.no>

MaochenWang1 commented on subsys/net/lib/wifi_credentials/wifi_credentials_shell.c:
```diff
@@ -309,8 +311,33 @@ static int cmd_auto_connect(const struct shell *sh, size_t argc, char *argv[])
+		      "0:None, 1:WPA2-PSK, 2:WPA2-PSK-256, 3:SAE-HNP, 4:SAE-H2E, 5:SAE-AUTO, 6:WAPI,"
+		      " 7:EAP-TLS, 8:WEP, 9: WPA-PSK, 10: WPA-Auto-Personal, 11: DPP\n"
+		      "12: EAP-PEAP-MSCHAPv2, 13: EAP-PEAP-GTC, 14: EAP-TTLS-MSCHAPv2,\n"
+		      "15: EAP-PEAP-TLS, 16:EAP_TLS_SHA256\n"
```
EAP_TLS_SHA256 is not used any more.
Please refer to https://github.com/zephyrproject-rtos/zephyr/blob/8473928db4e578704758c9d1c9669580902207c6/subsys/net/l2/wifi/wifi_shell.c#L3685

MaochenWang1 commented on subsys/net/lib/wifi_credentials/wifi_credentials_shell.c:
```diff
@@ -309,8 +311,33 @@ static int cmd_auto_connect(const struct shell *sh, size_t argc, char *argv[])
+		      "[-a, --anon-id]: Anonymous identity for enterprise mode.\n"
+		      "[-K, --key1-pwd for eap phase1 or --key2-pwd for eap phase2]:\n"
+		      "Private key passwd for enterprise mode. Default no password for private key.\n"
+		      "[-S, --suiteb-type]: 1:suiteb, 2:suiteb-192. Default 0: not suiteb mode.\n"
```
please refer to https://github.com/zephyrproject-rtos/zephyr/blob/8473928db4e578704758c9d1c9669580902207c6/subsys/net/l2/wifi/wifi_shell.c#L3693

krish2718 commented on include/zephyr/net/wifi_credentials.h:
```diff
@@ -56,6 +62,10 @@ struct wifi_credentials_header {
 	uint32_t timeout;                 /**< Timeout for connecting to the network */
 	uint8_t bssid[WIFI_MAC_ADDR_LEN]; /**< BSSID (Basic Service Set Identifier) */
 	uint8_t channel;                  /**< Channel on which the network operates */
+	char anon_id[16];                 /**< Anonymous identifier */
```
IIRC, 16 was a hack that we did to WAR the length limitations of Wi-Fi credentials backends, can you please revisit this? At a minimum we need to document this limitation somewhere.

D-Triveni replied:
For the time being, I have documented the `anon_id` limitation with the current value being used.

krish2718 commented on include/zephyr/net/wifi_credentials.h:
```diff
@@ -56,6 +62,10 @@ struct wifi_credentials_header {
 	uint8_t channel;                  /**< Channel on which the network operates */
+	char anon_id[16];                 /**< Anonymous identifier */
+	uint8_t aid_length;               /**< Length of the Anonymous identifier */
+	char key_passwd[16];              /**< Password/PSK */
```
IIRC, 16 was a hack that we did to WAR the length limitations of Wi-Fi credentials backends, can you please revisit this? At a minimum we need to document this limitation somewhere.

D-Triveni replied:
For the time being, I have documented the `key_passwd` limitation with the current value being used.

krish2718 commented on include/zephyr/net/wifi_credentials.h:
```diff
@@ -56,6 +62,18 @@ struct wifi_credentials_header {
 	uint8_t bssid[WIFI_MAC_ADDR_LEN]; /**< BSSID (Basic Service Set Identifier) */
 	uint8_t channel;                  /**< Channel on which the network operates */
+	/**
+	 * < Anonymous identifier
```
are you using `<` as a special identifier?

jukkar replied:
The `<` character does not work here, it tells doxygen that the variable we are commenting is on left side of the comment which is not the case here.

jukkar commented on include/zephyr/net/wifi_credentials.h:
```diff
@@ -56,6 +62,18 @@ struct wifi_credentials_header {
+	char anon_id[16];
+	uint8_t aid_length;               /**< Length of the Anonymous identifier */
+	/**
+	 * < Password/PSK
```
same here

krish2718 commented on include/zephyr/net/wifi_credentials.h:
```diff
@@ -56,6 +62,16 @@ struct wifi_credentials_header {
+	/** Anonymous identifier (Limited to 16 bytes due to settings subsystem overflow) */
+	char anon_id[16];
+
+	uint8_t aid_length;               /**< Length of the Anonymous identifier */
```
this is still old format, I would first have a commit that fixes all docstrings to be compliant to Zephyr and then do these changes.

D-Triveni committed:
include: zephyr: net: Fix docstrings in wifi_credentials header

Fix all docstrings in header to ensure compliance
with Zephyr's documentation guidelines.

Signed-off-by: Triveni Danda <triveni.danda@nordicsemi.no>

krish2718 reviewed (commented):
For Enterprise connection we also need certificates which are not passed through CLI instead they are passed build-time in `Wi-Fi shell`, so, we need to replicate the same here as well (that is what we did in the NCS before upstreaming of this module), or find some way to reuse them

D-Triveni committed:
wifi: shell: Add support for EAP-TLS method

Add support to read identity and private key password if
configured in Enterprise mode.

Signed-off-by: Triveni Danda <triveni.danda@nordicsemi.no>

D-Triveni commented:
> For Enterprise connection we also need certificates which are not passed through CLI instead they are passed build-time in `Wi-Fi shell`, so, we need to replicate the same here as well (that is what we did in the NCS before upstreaming of this module), or find some way to reuse them

Addressed.

krish2718 commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -1455,11 +1455,28 @@ BUILD_ASSERT(sizeof(CONFIG_WIFI_CREDENTIALS_STATIC_SSID) != 1,
 	     "CONFIG_WIFI_CREDENTIALS_STATIC_SSID required");
 #endif /* defined(CONFIG_WIFI_CREDENTIALS_STATIC) */
 
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_CRYPTO_ENTERPRISE
+static const char ca_cert_test[] = {
+#include <wifi_enterprise_test_certs/ca.pem.inc>
+	'\0'};
+
+static const char client_cert_test[] = {
+#include <wifi_enterprise_test_certs/client.pem.inc>
+	'\0'};
+
+static const char client_key_test[] = {
+#include <wifi_enterprise_test_certs/client-key.pem.inc>
+	'\0'};
+#endif /* CONFIG_WIFI_NM_WPA_SUPPLICANT_CRYPTO_ENTERPRISE */
+
```
these don't belong in the Wi-Fi management, these have to come from one layer above Wi-Fi management e.g., Wi-Fi shell or Wi-Fi cred of Customer Application. Wi-Fi management just provides `NET_REQUEST_WIFI_ENTERPRISE_CREDS`.

Before upstreaming, these certs reside in Wi-Fi credentials shell.
